### PR TITLE
use pmap to normalize reward model

### PR DESCRIPTION
**Pull request by liutianlin0121**
Hey!

As discussed in our [previous PR](https://github.com/vwxyzjn/lm-human-preference-details/pull/13), we can use `pmap` to accelerate the reward model normalization step. This PR does that. 

Fig below shows that the run time comparison between the pmapped version (cyan) and the previous version (grey). Both versions use `normalize_before` and `normalize_after`. The normalization step of the pmapped version is faster.

Screenshot from 2023-09-02 11-41-21

Test results are almost identical:
Screenshot from 2023-09-02 11-43-54

Additionally, I noted that in our previous pytorch and jax versions, we have an argument `args.normalize_samples` not used. We instead used `args.local_normalize_samples` to effectively mean `args.normalize_samples`. In this PR, I replaced all `args.local_normalize_samples` by `args.normalize_samples`. But feel free to suggest edits if `args.local_normalize_samples` is still needed.

### CODE REVIEW

**Comment by vwxyzjn** [comment]
Nice PR! Please keep the `normalize_samples`. We should calculate it as `normalize_samples =  int(args.local_normalize_samples * len(local_devices) * args.world_size)`

**Comment by liutianlin0121** [comment]
>keep the normalize_samples. We should calculate it as normalize_samples =  int(args.local_normalize_samples * len(local_devices) * args.world_size)

I see! Done.

File: lm_human_preference_details/train_reward_jax.py [hunk 1]
```diff
@@ -402,11 +401,21 @@ def normalize(
     generation_config,
 ):
     # number of minibatches for computing the normalization statistics
-    n_batches = ceil_div(args.local_normalize_samples, args.rollout_batch_size)
+    n_batches = ceil_div(args.normalize_samples, args.rollout_batch_size)
```
Comment (vwxyzjn) [inline lm_human_preference_details/train_reward_jax.py]: This should still be `local_normalize_samples`, right?

**Review by vwxyzjn — COMMENTED** [review]
Quick comment and everything else LG

File: lm_human_preference_details/train_reward_jax.py (hunk 1 above)
Comment (liutianlin0121) [reply lm_human_preference_details/train_reward_jax.py]: Thanks! I think `ceil_div(args.normalize_samples, args.rollout_batch_size)` can be correct? But let me know if I've overlooked anything:

In the jax version the dataloder `normalization_dataloader` returns `args.rollout_batch_size` samples per iteration. The `args.rollout_batch_size` is the number of samples for all devices. For this reason, if we want the model to see `args.normalize_samples` samples in total from all devices, we need
`ceil_div(args.normalize_samples, args.rollout_batch_size)` batches.

The situation is different in `train_reward_accelerate` though. There, we do `normalization_dataloader = accelerator.prepare(normalization_dataloader)`. In that case, I think the `args.rollout_batch_size` argument of dataloader refers to the number of samples per device. So we set `n_batches = ceil_div(args.local_normalize_samples, args.rollout_batch_size)` in that case.

Both versions should be equivalent now, although the per-device batch size is different. If we want to collect normalization statistics with the same batch size per device, perhaps we can define a `local_rollout_batch_size`? I think that with `local_rollout_batch_size`, we don't need `args.local_normalize_samples` anymore.

Comment (vwxyzjn) [reply lm_human_preference_details/train_reward_jax.py]: We should add a `args.rollout_batch_size` and a `args.local_rollout_batch_size`. 

PT: `rollout_batch_size = world_size * 1 * local_rollout_batch_size`
JAX: `rollout_batch_size = world_size * len(devices) * local_rollout_batch_size`

So if we want to align two settings to have `rollout_batch_size  = 8` with 8 GPUs, we need to set `local_rollout_batch_size=1` in PT. In jax, we need to set `local_rollout_batch_size=1` and `dataloader(batch_size=local_rollout_batch_size * len(devices)` because we shard the batch later.


Btw jax version can be made equivalent to pytorch with 


```
# world_size = 1, len(devices) = 8
WANDB_TAGS="tf_adam,gpt2" python benchmark/benchmark.py \
    --command "python lm_human_preference_details/train_reward_jax.py" \
    --num-seeds 10 \
    --start-seed 1 \
    --workers 2 \
    --slurm-gpus-per-task 8 \
    --slurm-ntasks 1 \
    --slurm-total-cpus 64 \
    --slurm-template-path benchmark/trl.slurm_template
```
```diff
# world_size = 8, len(devices) = 1
WANDB_TAGS="tf_adam,gpt2" python benchmark/benchmark.py \
+    --command "python lm_human_preference_details/train_reward_jax.py --distributed" \
    --num-seeds 10 \
    --start-seed 1 \
    --workers 2 \
+    --slurm-gpus-per-task 1 \
+    --slurm-ntasks 8 \
    --slurm-total-cpus 64 \
    --slurm-template-path benchmark/trl.slurm_template
```

Comment (liutianlin0121) [reply lm_human_preference_details/train_reward_jax.py]: Yep, just added `args.local_rollout_batch_size`. Both versions should be consistent as before. The difference at the implementation level is the `batch_size` we pass to the dataloaders:

- in pytorch:
```python
normalization_dataloader = DataLoader(normalization_dataset, batch_size=args.local_rollout_batch_size)
reward_model, optimizer, normalization_dataloader = accelerator.prepare(reward_model, optimizer, normalization_dataloader)
```

- in jax:
```python
normalization_dataloader = DataLoader(normalization_dataset, batch_size=args.rollout_batch_size)
```

**Comment by vwxyzjn** [comment]
Per OAI's setting, I think `local_rollout_batch_size = 512`. I also added the calculation.

**Comment by vwxyzjn** [comment]
Added two quick fixes. @liutianlin0121 let me know if they look good to you. If so feel free to merge.

File: lm_human_preference_details/train_reward_jax.py [hunk 2]
```diff
@@ -636,7 +649,7 @@ def train(args: Args):
         start_text=args.task.start_text,
         end_text=args.task.end_text,
     )
-    normalization_dataloader = DataLoader(normalization_dataset, batch_size=args.rollout_batch_size)
+    normalization_dataloader = DataLoader(normalization_dataset, batch_size=args.rollout_batch_size * len(local_devices))
```
Comment (liutianlin0121) [inline lm_human_preference_details/train_reward_jax.py]: Here I think it should be just `batch_size=args.rollout_batch_size`, because `args.rollout_batch_size` is already the global batch size scaled by the number of devices: `args.rollout_batch_size = int(args.local_rollout_batch_size * len(local_devices) * args.world_size)` in line 588. Right?

Comment (vwxyzjn) [reply lm_human_preference_details/train_reward_jax.py]: Imagine we have 

* `rollout_batch_size = 16`
* `local_rollout_batch_size = 1`
* `len(devices) = 8` (8 GPUs)
* `world_size=2` (2 distributed training processes, 16 GPUs total)

Then the data loader should have a batch size of 8, which we then shard to 8 devices locally in each of the two processes.

Comment (vwxyzjn) [reply lm_human_preference_details/train_reward_jax.py]: Oh sorry, it should be `batch_size=args.local_rollout_batch_size * len(local_devices)`. Just added the fix.

Comment (liutianlin0121) [reply lm_human_preference_details/train_reward_jax.py]: I see, thanks!
